# Domaći zadatak 2: Genomska Informatika

### Teodora Radaljac 3098/2023

Instaliranje **pysam** biblioteke

In [26]:
pip install pysam

In [27]:
import pysam

In [28]:
path_to_file = 'merged-tumor.bam'

# creating an alignment file for merged-tumor.bam
bamfile = pysam.AlignmentFile(path_to_file, 'rb')

In [29]:
# taking the first read
first_read = next(bamfile)

print(first_read)
print(first_read.flag)

C0HVYACXX120402:7:1207:5722:57044	1187	#20	9483249	27	76M	#20	9483382	209	TTTTCAAACAGTATCTATGCCTGCCAAATGTGAACATATAAAAAAAAACCAGAATGTGCCATTCTGATTTAAACTG	array('B', [28, 28, 27, 29, 31, 30, 31, 31, 29, 31, 35, 30, 29, 31, 34, 30, 29, 23, 41, 32, 20, 30, 29, 34, 34, 29, 30, 31, 30, 30, 30, 33, 33, 26, 39, 12, 25, 19, 32, 30, 35, 28, 35, 33, 23, 33, 35, 36, 30, 38, 33, 41, 34, 35, 31, 33, 23, 30, 30, 36, 27, 32, 29, 34, 35, 41, 33, 31, 33, 29, 32, 32, 31, 31, 31, 34])	[('XA', 'GL000217.1,-110754,76M,1;'), ('BD', 'NNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNN'), ('MD', '76'), ('RG', '1'), ('BI', 'NNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNN'), ('NM', 0), ('MQ', 27), ('AS', 76), ('XS', 71)]
1187


Ispisom first read-a možemo videti vrednosti polja u njemu.

Redom:
> Query name: C0HVYACXX120402:7:1207:5722:57044


> Bitwise flag: 1187

> Reference sequence name: #20

> POS (starting position of the alignment in the reference sequence): 9483249

> MAPping Quality: 27

> CIGAR string (how the read aligns to the reference): 76M

> Reference name of the mate/next read: #20

> Position of the mate/next read: 9483382

> Template length: 209

> Segment sequence: TTTTCAAACAGTATCTATGCCTGCCAAATGTGAACATATAAAAAAAAACCAGAATGTGCCATTCTGATTTAAACTG

> Quality scores: niz sa puno elemenata

Takodje možemo proveriti i koji su flag-ovi aktivni a koji ne.

Aktivni:

>   read paired
  
>  read mapped in proper pair

>  mate reverse strand

>  second in pair

>  read is PCR or optical duplicate

Neaktivni:

>   read unmapped
  
> mate unmapped

>   read reverse strand

> first in pair

> not primary alignment

>   read fails platform/vendor quality checks

> supplementary alignment

In [30]:
total_reads = 0
unmapped_reads = 0
reads_quality_zero = 0
total_quality = 0

Prolazak kroz sve read-ove i prebrojavanje i izračunavanje odgovarajućih karakteristika.

In [31]:
bamfile = pysam.AlignmentFile(path_to_file, 'rb')

for read in bamfile:
    total_reads += 1
    if read.is_unmapped:
        unmapped_reads += 1
    if read.mapping_quality == 0:
        reads_quality_zero += 1
    total_quality += read.mapping_quality

average_quality = total_quality / total_reads if total_reads else 0
average_quality_excluding_zero = (total_quality - reads_quality_zero) / (total_reads - reads_quality_zero) if (total_reads - reads_quality_zero) else 0

Ispis traženih vrednosti.

In [32]:
print("Total reads: ", total_reads)
print("Unmapped reads: ", unmapped_reads)
print("Reads with quality 0: ", reads_quality_zero)
print("Average mapping quality for all reads: ", average_quality)
print("Average mapping quality excluding reads with quality 0: ", average_quality_excluding_zero)


Total reads:  2921629
Unmapped reads:  17765
Reads with quality 0:  126628
Average mapping quality for all reads:  55.91379158681681
Average mapping quality excluding reads with quality 0:  58.40167033929505


In [33]:
bamfile.close()